packages/mds-compliance-service/repository/index.ts

```diff
@@ -0,0 +1,175 @@
+import { UUID } from '@mds-core/mds-types'
+import {
+  ComplianceSnapshotDomainModel,
+  GetComplianceSnapshotsByTimeIntervalOptions,
+  GetComplianceSnapshotOptions
+} from '../@types'
+import { ComplianceSnapshotEntityToDomain, ComplianceSnapshotDomainToEntityCreate } from './mappers'
+import { ComplianceSnapshotEntity, ComplianceSnapshotEntityModel } from './entities/compliance-snapshot-entity'
+import migrations from './migrations'
+
+class ComplianceSnapshotReadWriteRepository extends ReadWriteRepository {
+  public getComplianceSnapshot = async ({
+    compliance_snapshot_id,
+    provider_id,
+    policy_id,
+    compliance_as_of = now()
+  }: GetComplianceSnapshotOptions): Promise<ComplianceSnapshotDomainModel> => {
+    const { connect } = this
+    try {
+      const connection = await connect('ro')
+      if (isDefined(compliance_snapshot_id)) {
+        const entity = await connection.getRepository(ComplianceSnapshotEntity).findOne({
+          where: {
+            compliance_snapshot_id
+          }
+        })
+        if (!entity) {
+          throw new NotFoundError(`ComplianceSnapshot ${compliance_snapshot_id} not found`)
+        }
+        return ComplianceSnapshotEntityToDomain.map(entity)
+      }
+      if (!isDefined(provider_id) || !isDefined(policy_id)) {
+        throw RepositoryError('provider_id and policy_id must be given if compliance_snapshot_id is not given')
+      }
+
+      const query = connection
+        .getRepository(ComplianceSnapshotEntity)
+        .createQueryBuilder()
+        .where(`provider_id = '${provider_id}'`)
+        .andWhere(`policy_id = '${policy_id}'`)
+        .andWhere(`compliance_as_of >= ${compliance_as_of}`)
+        .orderBy('compliance_as_of')
+
+      const entity = await query.getOne()
+      if (!entity) {
+        throw new NotFoundError(
+          `ComplianceSnapshot not found with params ${JSON.stringify({
+            policy_id,
+            provider_id,
+            compliance_as_of
+          })} not found`
+        )
+      }
+      return ComplianceSnapshotEntityToDomain.map(entity)
+    } catch (error) {
+      throw RepositoryError(error)
+    }
+  }
+
+  public getComplianceSnapshotsByTimeInterval = async ({
+    start_time,
+    end_time = now(),
+    provider_ids,
+    policy_ids
+  }: GetComplianceSnapshotsByTimeIntervalOptions): Promise<ComplianceSnapshotDomainModel[]> => {
+    const { connect } = this
+    try {
+      const connection = await connect('ro')
+      const query = connection
+        .getRepository(ComplianceSnapshotEntity)
+        .createQueryBuilder()
+        .where(`compliance_as_of >= ${start_time}`)
+        .andWhere(`compliance_as_of <= ${end_time}`)
+      if (isDefined(provider_ids)) {
+        query.andWhere('provider_id IN (:...provider_ids)', { provider_ids })
+      }
+      if (isDefined(policy_ids)) {
+        query.andWhere('policy_id IN (:...policy_ids)', { policy_ids })
+      }
```

Comment:
I wish there was some way to do conditional chained function calls inline, would be so fancy 😍 Like 
```typescript
const foo = someBar.func()(isDefined(provider_ids) ? .func(provider_ids) : undefined)(isDefined(policy_ids) ? .func(policy_ids) : undefined)
```